### ISSUE TITLE
[restrict-template-expressions] add option to allow regular expressions

### ISSUE THREAD

**Issue by JoshuaKGoldberg** [typescript-eslint/typescript-eslint#3671]
- [x] I have tried restarting my IDE and the issue persists.
- [x] I have updated to the latest version of the packages.
- [x] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
    "@typescript-eslint/restrict-template-expressions": ["error"]
  }
}
```

```TS
function log(value: RegExp) {
  console.log(`${value}`);
}
```

**Expected Result**

Since `RegExp` defines `.toString()`, I would expect that there would be options to allow printing them similar to `allowBoolean`.

**Actual Result**

No way to make this allowed without disabling the rule.

**Additional Info**

Context: hoping to enable this internally for #3278.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.28.5` |
| `@typescript-eslint/parser`        | `4.28.5` |
| `TypeScript`                       | `4.3.5` |
| `ESLint`                           | `7.27.0` |
| `node`                             | `16.6.0` |

**Comment by Ulrikop** [comment]
I would also appreciate an option to allow objects with a `toString` implementation (so not only for allowing RegExp).

I want to use the rule to avoid text like `[object Object]`, but I don't want to unnecessarily complicate the code in places where it is not necessary.
